src/gui/src/staGuiInterface.cpp

```diff
@@ -348,6 +349,313 @@ void TimingPath::setSlackOnPathNodes()
 
 /////////////
 
+ClockTree::ClockTree(ClockTree* parent)
+    : parent_(parent),
+      clock_(parent->clock_),
+      network_(parent->network_),
+      net_(nullptr),
+      level_(parent_->level_ + 1)
+{
+}
+
+ClockTree::ClockTree(sta::Clock* clock, sta::dbNetwork* network)
+    : parent_(nullptr),
+      clock_(clock),
+      network_(network),
+      net_(nullptr),
+      level_(0)
+{
+}
+
+std::set<sta::Pin*> ClockTree::getDrivers() const
+{
+  std::set<sta::Pin*> drivers;
+  for (const auto& [driver, arrival] : drivers_) {
+    drivers.insert(driver);
+  }
+  return drivers;
+}
+
+std::set<sta::Pin*> ClockTree::getLeaves() const
+{
+  std::set<sta::Pin*> leaves;
+  for (auto& [leaf, arrival] : leaves_) {
+    leaves.insert(leaf);
+  }
+  return leaves;
+}
+
+ClockTree* ClockTree::getTree(sta::Net* net)
+{
+  if (net == net_) {
+    return this;
+  }
+
+  for (const auto& fanout : fanout_) {
+    if (fanout->getNet() == net) {
+      return fanout.get();
+    }
+  }
+  ClockTree* tree = new ClockTree(this);
+  tree->setNet(net);
+  fanout_.emplace_back(tree);
+  return tree;
+}
+
+int ClockTree::getTotalLeaves() const
+{
+  int total = leaves_.size();
+
+  for (const auto& fanout : fanout_) {
+    total += fanout->getTotalLeaves();
+  }
+
+  return total;
+}
+
+int ClockTree::getTotalFanout() const
+{
+  int total = 0;
+  if (!leaves_.empty()) {
+    total = 1;
+  }
+
+  for (const auto& fanout : fanout_) {
+    total += fanout->getTotalFanout();
+  }
+
+  return total;
+}
+
+int ClockTree::getMaxLeaves() const
+{
+  int width = leaves_.size();
+
+  for (const auto& fanout : fanout_) {
+    width = std::max(width, fanout->getMaxLeaves());
+  }
+
+  return width;
+}
+
+sta::Delay ClockTree::getMinimumArrival() const
+{
+  sta::Delay minimum = std::numeric_limits<sta::Delay>::max();
+
+  for (const auto& [driver, arrival] : drivers_) {
+    minimum = std::min(minimum, arrival);
+  }
+
+  for (const auto& [leaf, arrival] : leaves_) {
+    minimum = std::min(minimum, arrival);
+  }
+
+  for (const auto& fanout : fanout_) {
+    minimum = std::min(minimum, fanout->getMinimumArrival());
+  }
+
+  return minimum;
+}
+
+sta::Delay ClockTree::getMaximumArrival() const
+{
+  sta::Delay maximum = std::numeric_limits<sta::Delay>::min();
+
+  for (const auto& [driver, arrival] : drivers_) {
+    maximum = std::max(maximum, arrival);
+  }
+
+  for (const auto& [leaf, arrival] : leaves_) {
+    maximum = std::max(maximum, arrival);
+  }
+
+  for (const auto& fanout : fanout_) {
+    maximum = std::max(maximum, fanout->getMaximumArrival());
+  }
+
+  return maximum;
+}
+
+sta::Delay ClockTree::getMinimumDriverDelay() const
+{
+  sta::Delay minimum = std::numeric_limits<sta::Delay>::max();
+
+  if (parent_ != nullptr) {
+    for (const auto& [driver, arrival] : drivers_) {
+      const auto other = parent_->getPairedReciever(driver);
+      minimum = std::min(minimum, arrival - other.second);
```

Comment:
As you used elsewhere:
 const auto& [receiver, time] = parent_->getPairedReciever(driver);